syntax = "proto3";

package shopping;

message Item {
  int32 id = 1;
  float price = 2;
  string name = 3;
  string category = 4;
  string description = 5;
  int32 quantity_remaining = 6;
  string seller_address = 7;
  float rating = 8;
}

message SellerRegistrationRequest {
  string seller_address = 1;
  string uuid = 2;
}

message SellerRegistrationResponse {
  string message = 1;
}

service Market {
  rpc RegisterSeller(SellerRegistrationRequest) returns (SellerRegistrationResponse);
  rpc SellItem(Item) returns (Item);
  rpc UpdateItem(Item) returns (Item);
  rpc DeleteItem(Item) returns (Item);
  rpc DisplaySellerItems(SellerRegistrationRequest) returns (stream Item);
}

message BuyerSearchRequest {
  string item_name = 1;
  string category = 2;
}

message BuyerBuyRequest {
  int32 item_id = 1;
  int32 quantity = 2;
  string buyer_address = 3;
}

message BuyerWishlistRequest {
  int32 item_id = 1;
  string buyer_address = 2;
}

message BuyerRateRequest {
  int32 item_id = 1;
  string buyer_address = 2;
  int32 rating = 3;
}

service Buyer {
  rpc SearchItem(BuyerSearchRequest) returns (stream Item);
  rpc BuyItem(BuyerBuyRequest) returns (Item);
  rpc AddToWishList(BuyerWishlistRequest) returns (Item);
  rpc RateItem(BuyerRateRequest) returns (Item);
}
